File: src/advent_of_code/2022/day01.clj
(ns advent-of-code.2022.day01
  "https://adventofcode.com/2022/day/1"
  (:require
   [clojure.string :as str]))

(def input-filename "inputs/2022/day01.txt")

(def total-calories-carried-by-each-elf
  (->> (slurp input-filename)
       (str/split-lines)
       (partition-by str/blank?)
       (remove #(= % '("")))
       ;; we now have (("4887" "9307") ("8895" "8136" "6292" "10177" "4077" "8228") ...)
       (map (fn [list-cal-strs]
              (->> (map #(Integer/parseInt %) list-cal-strs)
                   (reduce +))))))

(def part-1-answer
  "Maximum number of calories carried by any elf"
  (->> total-calories-carried-by-each-elf
       (reduce max)))

(def part-2-answer
  "Sum of calories carried by the 3 elves carrying the 3 highest number of calories"
  (->> total-calories-carried-by-each-elf
       sort
       (take-last 3)
       (reduce +)))

(defn -main []
  (println "part-1 answer:" part-1-answer)
  (println "part-2 answer:" part-2-answer))

(comment
  (-main)
  )

File: python/2022/day01.py
"""Day 1: Calorie Counting - https://adventofcode.com/2022/day/1"""

INPUT_FILE = "inputs/2022/day01.txt"


def parse_input(filename: str) -> list[list[int]]:
    """Parse input into groups of calories per elf."""
    with open(filename) as f:
        content = f.read().strip()

    groups = content.split("\n\n")
    return [[int(cal) for cal in group.split("\n")] for group in groups]


def total_calories_per_elf(elf_groups: list[list[int]]) -> list[int]:
    """Calculate total calories carried by each elf."""
    return [sum(group) for group in elf_groups]


def part1(totals: list[int]) -> int:
    """Maximum number of calories carried by any elf."""
    return max(totals)


def part2(totals: list[int]) -> int:
    """Sum of calories carried by the 3 elves with most calories."""
    return sum(sorted(totals, reverse=True)[:3])


def main():
    elf_groups = parse_input(INPUT_FILE)
    totals = total_calories_per_elf(elf_groups)

    answer1 = part1(totals)
    answer2 = part2(totals)

    print(f"Part 1: {answer1}")
    print(f"Part 2: {answer2}")

    return answer1, answer2


if __name__ == "__main__":
    main()
